joblib/memory.py

```diff
@@ -802,14 +796,20 @@ class Memory(Logger):
         backend_options: dict, optional
             Contains a dictionnary of named parameters used to configure
             the store backend.
+
+        cachedir: str or None, optional
+
+            .. deprecated: 0.12
+                'cachedir' has been deprecated in 0.12 and will be
+                removed in 0.14. Use the 'location' parameter instead.
     """
     # ------------------------------------------------------------------------
     # Public interface
     # ------------------------------------------------------------------------
 
-    def __init__(self, location=None, backend='local', cachedir=None,
-                 mmap_mode=None, compress=False, verbose=1, bytes_limit=None,
-                 backend_options={}):
+    def __init__(self, location=None, backend='local', mmap_mode=None,
```

Comment:
I think the problem is more general than this. I reckon the fix is to keep the arguments order as they were in 0.11 otherwise we will break Memory pickles between 0.11 and 0.12 with no particularly good reason.

scikit-learn 0.19 has a memory argument in Pipeline so this could affect pickled Pipelines between scikit-learn 0.19 and 0.20.

With joblib 0.11:
```py
import pickle

from joblib import __version__
from joblib import Memory

mem = Memory(cachedir='/tmp/test')

print(pickle.dumps(mem)) 
# b'\x80\x03cjoblib.memory\nMemory\nq\x00(X\t\x00\x00\x00/tmp/testq\x01N\x89K\x01tq\x02Rq\x03.'
```

Reloading in 0.12+ (even with the current fix from this PR):
```py
import pickle
pickle.loads(b'\x80\x03cjoblib.memory\nMemory\nq\x00(X\t\x00\x00\x00/tmp/testq\x01N\x89K\x01tq\x02Rq\x03.')
```

```py-tb
/home/lesteve/miniconda3/bin/ipython:2: UserWarning: Compressed results cannot be memmapped
  
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-926293d23d86> in <module>()
      1 import pickle
----> 2 pickle.loads(b'\x80\x03cjoblib.memory\nMemory\nq\x00(X\t\x00\x00\x00/tmp/testq\x01N\x89K\x01tq\x02Rq\x03.')

~/dev/joblib/joblib/memory.py in __init__(self, location, backend, mmap_mode, compress, verbose, bytes_limit, backend_options, cachedir)
    845             backend, location, verbose=self._verbose,
    846             backend_options=dict(compress=compress, mmap_mode=mmap_mode,
--> 847                                  **backend_options))
    848 
    849     @property

~/dev/joblib/joblib/memory.py in _store_backend_factory(backend, location, verbose, backend_options)
    115         if obj is None:
    116             raise TypeError('Unknown location {0} or backend {1}'.format(
--> 117                             location, backend))
    118 
    119         # The store backend is configured with the extra named parameters,

TypeError: Unknown location /tmp/test/joblib or backend None
```

Reply:
> I reckon the fix is to keep the arguments order as they were in 0.11 otherwise we will break Memory pickles between 0.11 and 0.12 with no particularly good reason.

I guess actually you should keep `location` (rather than `cachedir`) as the first argument.

Reply:
We should probably change `__reduce__` to use arguments via a dictionary rather than positionally to avoid this kind of hassles in the future ...